File: backend/rust/src/domain/value/email.rs
//! Email value object.
//! Keeps validation in one place and guarantees: if you have an `Email`, it's valid.

use core::fmt;
use core::str::FromStr;

use crate::domain::errors::DomainError;

#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct Email(String);

impl Email {
    pub fn as_str(&self) -> &str {
        &self.0
    }

    pub fn into_string(self) -> String {
        self.0
    }
}

impl fmt::Display for Email {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(&self.0)
    }
}

impl FromStr for Email {
    type Err = DomainError;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let s = s.trim();
        if s.is_empty() {
            return Err(DomainError::validation("Email must not be empty"));
        }

        let parts: Vec<&str> = s.split('@').collect();
        if parts.len() != 2 {
            return Err(DomainError::validation("Email must contain exactly one '@'"));
        }

        let local = parts[0];
        let domain = parts[1];

        if local.is_empty() || domain.is_empty() {
            return Err(DomainError::validation("Email local/domain must not be empty"));
        }
        if !domain.contains('.') {
            return Err(DomainError::validation("Email domain must contain a '.'"));
        }

        Ok(Email(s.to_string()))
    }
}

impl TryFrom<String> for Email {
    type Error = DomainError;

    fn try_from(value: String) -> Result<Self, Self::Error> {
        Email::from_str(&value)
    }
}

impl From<Email> for String {
    fn from(value: Email) -> Self {
        value.0
    }
}

File: backend/c/domain/value/email.c
#include "../include/email.h"
#include <string.h>

const char *email_as_cstr(const Email *email) {
    return email ? email->value : "";
}

bool email_try_create(const char *raw, Email *out, DomainError *err) {
    if (!raw || !out) {
        if (err) *err = domain_error_validation("Email must not be null");
        return false;
    }

    /* trim left/right */
    const char *start = raw;
    while (*start == ' ' || *start == '\t' || *start == '\n' || *start == '\r')
        ++start;

    const char *end = start + strlen(start);
    while (end > start && (end[-1] == ' ' || end[-1] == '\t' ||
                           end[-1] == '\n' || end[-1] == '\r')) {
        --end;
                           }

    size_t len = (size_t)(end - start);
    if (len == 0) {
        if (err) *err = domain_error_validation("Email must not be empty");
        return false;
    }
    if (len > EMAIL_MAX_LEN) {
        if (err) *err = domain_error_validation("Email too long");
        return false;
    }

    char buffer[EMAIL_MAX_LEN + 1];
    memcpy(buffer, start, len);
    buffer[len] = '\0';

    char *at = strchr(buffer, '@');
    if (!at || strchr(at + 1, '@')) {
        if (err) *err =
                domain_error_validation("Email must contain exactly one '@'");
        return false;
    }

    char *local = buffer;
    char *domain = at + 1;
    *at = '\0';

    if (*local == '\0' || *domain == '\0') {
        if (err) *err = domain_error_validation(
                "Email local/domain must not be empty");
        return false;
    }

    if (!strchr(domain, '.')) {
        if (err) *err =
                domain_error_validation("Email domain must contain a '.'");
        return false;
    }

    *at = '@';

    strcpy(out->value, buffer);
    if (err) *err = domain_error_ok();
    return true;
}
